lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+/**
+ * A metrics provider which reports to diagnostics_channel.
+ *
+ * # Metric Types
+ *
+ * - Counter: An increasing or decreasing value.
+ * - Gauge: A snapshot of a single value in time.
+ * - Meter: A number of events per interval.
+ * - Timer: A duration in milliseconds.
+ * - UniqueSet: A unique count of number of unique values seen.
+ * - PeriodicGauge: A gauge which periodically updates its value by calling a function.
+ *
+ * # Exporting Metrics
+ *
+ * Several text format exporters are provided as streams:
+ * - Statsd
+ * - Dogstatsd
+ * - Graphite
+ * - Prometheus
+ *
+ * # TODO(qard):
+ * - Histograms
+ * - Distributions/Summaries
+ */
+
+'use strict';
+
+const {
+  ArrayPrototypeJoin,
+  ArrayPrototypeMap,
+  MathFloor,
+  ObjectAssign,
+  ObjectEntries,
+  ObjectFreeze,
+  ObjectKeys,
+  SafeMap,
+  SafeWeakSet,
+  SymbolDispose,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { setInterval, clearInterval } = require('internal/timers');
+
+const {
+  channel,
+  hasChannel,
+  subscribe,
+  unsubscribe,
+} = require('diagnostics_channel');
+const { performance } = require('perf_hooks');
+const { Transform } = require('stream');
+
+const newMetricChannel = channel('metrics:new');
+
+/**
+ * Mix two metadata objects together.
+ * @param {object} a The first metadata object.
+ * @param {object} b The second metadata object.
+ * @returns {object} The mixed metadata.
+ * @private
+ */
+function mixMeta(a, b) {
+  if (a === undefined) return b;
+  if (b === undefined) return a;
+  return ObjectAssign({}, a, b);
+}
+
+/**
+ * Represents a single reported metric.
+ */
+class MetricReport {
+  /**
+   * The type of metric.
+   * @property {string} type
+   */
+
+  /**
+   * The name of the metric.
+   * @property {string} name
+   */
+
+  /**
+   * The value of the metric.
+   * @property {number} value
+   */
+
+  /**
+   * Additional metadata to include with the report.
+   * @property {object} meta
+   */
+
+  /**
+   * Constructs a new metric report.
+   * @param {string} type The type of metric.
+   * @param {string} name The name of the metric.
+   * @param {number} value The value of the metric.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  constructor(type, name, value, meta) {
+    this.type = type;
+    this.name = name;
+    this.value = value;
+    this.meta = meta;
+    this.time = performance.now();
+    ObjectFreeze(this);
```

Comment:
why freezing?

Reply:
Exposing these as getters to on the prototype would achieve immutability on the defined members. I don't really see much reason for freezing the entire object.